### PR TITLE
Counter Implementation

### PR DESCRIPTION
**Pull request by Ja-ake**

### CODE REVIEW

File: CommonChickenRuntimeEngine/src/ccre/frc/FRC.java [hunk 1]
```diff
@@ -69,6 +69,34 @@
...
+
+        /**
+         * The integer value representing this enumeration
+         */
+        public final int value;
+
+        private CounterMode(int value) {
+            this.value = value;
+        }
+    }
```
Comment (col6y) [inline CommonChickenRuntimeEngine/src/ccre/frc/FRC.java]: Is there any good way that we could do this without introducing another enum?

File: CommonChickenRuntimeEngine/src/ccre/frc/FRC.java [hunk 2]
```diff
@@ -723,6 +751,79 @@ public static FloatInput makeEncoder(int aChannel, int bChannel, boolean reverse
...
+     * Creates a reference to a Counter on the specified ports* with the
+     * specified number of average bits. Will count up when up channel
+     * is pulsed, and down when down channel is pulsed.
+     * 
+     * @param channel The channel for the counter.
+     * @param resetWhen If provided, the Counter's value will be reset when
+     * this event is produced.
+     * @param updateOn Updates the FloatInput when this event is produced.
+     * @return the Counter, reporting Counter ticks.
+     */
```
Comment (col6y) [inline CommonChickenRuntimeEngine/src/ccre/frc/FRC.java]: "Average bits"? Also, you need both channels specified in the Javadoc.

File: CommonChickenRuntimeEngine/src/ccre/frc/FRC.java [hunk 3]
```diff
@@ -723,6 +751,79 @@ public static FloatInput makeEncoder(int aChannel, int bChannel, boolean reverse
...
+    public static FloatInput makeCounter(int upChannel, int downChannel, EventInput resetWhen, EventInput updateOn, CounterMode mode) {
+        // This exception must be thrown only for downChannel as upChannel 
+        // is checked in the implementation. If the implementation is supplied
+        // a negative down channel, it will ignore the down channel which should
+        // only be used in the other function
+        if (downChannel < 0) throw new RuntimeException("Invalid down channel: " + downChannel);
```
Comment (col6y) [inline CommonChickenRuntimeEngine/src/ccre/frc/FRC.java]: Please format your contributions.

File: CommonChickenRuntimeEngine/src/ccre/frc/FRC.java [hunk 4]
```diff
@@ -723,6 +751,79 @@ public static FloatInput makeEncoder(int aChannel, int bChannel, boolean reverse
...
+     * @param updateOn Updates the FloatInput when this event is produced.
+     * @return the Counter, reporting Counter ticks.
+     */
+    public static FloatInput makeCounter(int upChannel, int downChannel, EventInput resetWhen, EventInput updateOn) {
+        // This exception must be thrown only for downChannel as upChannel 
+        // is checked in the implementation. If the implementation is supplied
+        // a negative down channel, it will ignore the down channel which should
+        // only be used in the other function
+        if (downChannel < 0) throw new RuntimeException("Invalid down channel: " + downChannel);
+        return impl.makeCounter(upChannel, downChannel, resetWhen, sensorPeriodic, CounterMode.kTwoPulse);
```
Comment (col6y) [inline CommonChickenRuntimeEngine/src/ccre/frc/FRC.java]: Why is `updateOn` not used?

File: CommonChickenRuntimeEngine/src/ccre/frc/FRC.java [hunk 5]
```diff
@@ -723,6 +751,79 @@ public static FloatInput makeEncoder(int aChannel, int bChannel, boolean reverse
...
+     * @return the Counter, reporting Counter ticks.
+     */
+    public static FloatInput makeCounter(int upChannel, int downChannel, EventInput resetWhen) {
+        // This exception must be thrown only for downChannel as upChannel 
+        // is checked in the implementation. If the implementation is supplied
+        // a negative down channel, it will ignore the down channel which should
+        // only be used in the other function
+        if (downChannel < 0) throw new RuntimeException("Invalid down channel: " + downChannel);
+        return impl.makeCounter(upChannel, downChannel, resetWhen, sensorPeriodic, CounterMode.kTwoPulse);
+    }
```
Comment (col6y) [inline CommonChickenRuntimeEngine/src/ccre/frc/FRC.java]: Better idea than specializing based on existence of `downChannel`: have a static constant (-1) on FRC that represents "this channel isn't used" and then have fewer methods.

File: CommonChickenRuntimeEngine/src/ccre/frc/FRCImplementation.java [hunk 6]
```diff
@@ -200,6 +200,22 @@
      */
     public FloatInput makeEncoder(int aChannel, int bChannel, boolean reverse, EventInput resetWhen, EventInput updateOn);
 
+    
+    /**
+     * Creates a reference to a Counter on the specified ports* with the
+     * specified number of average bits.
+     * 
+     * <br />* If downChannel is negative, it will be ignored. Make sure
```
Comment (col6y) [inline CommonChickenRuntimeEngine/src/ccre/frc/FRCImplementation.java]: Why the stars, and why the `<br />`?

File: CommonChickenRuntimeEngine/src/ccre/frc/FRCImplementation.java [hunk 7]
```diff
@@ -200,6 +200,22 @@
...
+     * upChannel is a valid digital channel.
```
Comment (col6y) [inline CommonChickenRuntimeEngine/src/ccre/frc/FRCImplementation.java]: So - wait - is upChannel currently checked by the caller or callee here? Your docs and usage as a caller are different.

File: Emulator/src/ccre/frc/DeviceBasedImplementation.java [hunk 8]
```diff
@@ -496,4 +497,9 @@ public FloatInput addFloat(String name) {
     public EventInput getOnInitComplete() {
         return onInitComplete;
     }
+
+    @Override
+    public FloatInput makeCounter(int upChannel, int downChannel, EventInput resetWhen, EventInput updateOn, CounterMode mode) {
+        throw new UnsupportedOperationException("Not implemented.");
+    }
```
Comment (col6y) [inline Emulator/src/ccre/frc/DeviceBasedImplementation.java]: Implement this.

File: SampleIgneousRobot/bin/main.properties [hunk 9]
```diff
@@ -0,0 +1,4 @@
+igneous.main=org.team1540.minimal.Minimal
+remoteaddress-2014=10.15.40.2
+# To try out the new Direct launcher, uncomment this:
+#boot-class=ccre.igneous.direct.DirectIgneousLauncherImpl
```
Comment (col6y) [inline SampleIgneousRobot/bin/main.properties]: Please do NOT commit files in the `bin` directory! I will not merge this until none of the included commits have anything about this.

File: roboRIO/src/ccre/frc/DirectCounter.java [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+import java.nio.ByteOrder;
+import java.nio.IntBuffer;
+
+import ccre.log.Logger;
+import edu.wpi.first.wpilibj.hal.CounterJNI;
+import edu.wpi.first.wpilibj.hal.HALUtil;
+
+public class DirectCounter {
+
+    private static int COUNTER_NUM = 26;  // ???? who even knows
```
Comment (col6y) [inline roboRIO/src/ccre/frc/DirectCounter.java]: Try to figure this out.

File: roboRIO/src/ccre/frc/DirectCounter.java [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+        public static final EncodingType k2X = new EncodingType(k2X_val);
+        /**
+         * Count rising and falling on both channels
+         */
+        public static final EncodingType k4X = new EncodingType(k4X_val);
+
+        private EncodingType(int value) {
+            this.value = value;
+        }
+    }
```
Comment (col6y) [inline roboRIO/src/ccre/frc/DirectCounter.java]: You never use this.

File: roboRIO/src/ccre/frc/DirectCounter.java [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+
+        /**
+         * The integer value representing this enumeration
+         */
+        public final int value;
+
+        private Mode(int value) {
+            this.value = value;
+        }
+    }
```
Comment (col6y) [inline roboRIO/src/ccre/frc/DirectCounter.java]: Do you *really* need an enum?

File: roboRIO/src/ccre/frc/DirectCounter.java [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+        Common.check(status);
+        
+        setUpSource(counter, channel);
+        
+        return counter;
+    }
+    
+    public static ByteBuffer init(int channelUp, int channelDown, Mode mode) {
+        IntBuffer status = Common.allocateInt();
+        IntBuffer index = Common.allocateInt();
```
Comment (col6y) [inline roboRIO/src/ccre/frc/DirectCounter.java]: Please use `getCheckBuffer()` for `status` as much as possible - it helps prevent overfilled heap.

File: roboRIO/src/ccre/frc/DirectCounter.java [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+        ByteBuffer counter = CounterJNI.initializeCounter(mode.value, index, status);
+        Common.check(status);
+        
+        setDownSource(counter, channelDown);
+        setUpSource(counter, channelUp);
+        
+        setUpSourceEdge(counter, true, false);
+        setDownSourceEdge(counter, false, true);
```
Comment (col6y) [inline roboRIO/src/ccre/frc/DirectCounter.java]: Why are the parameters like this? Please add a comment.

File: roboRIO/src/ccre/frc/DirectCounter.java [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+        
+        return counter;
+    }
+    
+    public static void setUpSource(ByteBuffer counter, int channel) {
+        ByteBuffer status = ByteBuffer.allocateDirect(4);
+        status.order(ByteOrder.LITTLE_ENDIAN);
```
Comment (col6y) [inline roboRIO/src/ccre/frc/DirectCounter.java]: Use Common.

File: roboRIO/src/ccre/frc/DirectDigital.java [hunk 16]
```diff
@@ -141,4 +141,13 @@ public static boolean waitForInterrupt(int id, float timeout, boolean ignorePrev
         Common.check(status);
         return irqs == 0;
     }
+    
+    public static ByteBuffer getDigitalSource(int channel) {
+        // if (digitals[channel] == null) throw new RuntimeException("Channel hasn't been initialized");
+        return digitals[channel];
```
Comment (col6y) [inline roboRIO/src/ccre/frc/DirectDigital.java]: Why did you remove this check?

File: roboRIO/src/ccre/frc/DirectFRCImplementation.java [hunk 17]
```diff
@@ -405,6 +405,35 @@ protected float apply() {
...
+    
+    @Override
+    public FloatInput makeCounter(int channelUp, int channelDown, EventInput resetWhen, EventInput updateOn, FRC.CounterMode mode) {
+        if (channelUp < 0) throw new RuntimeException("Invalid up channel: " + channelUp);
+        
+        DirectCounter.Mode counterMode;
+        if (mode == FRC.CounterMode.kExternalDirection) counterMode = DirectCounter.Mode.kExternalDirection;
+        else if (mode == FRC.CounterMode.kPulseLength) counterMode = DirectCounter.Mode.kPulseLength;
+        else if (mode == FRC.CounterMode.kSemiperiod) counterMode = DirectCounter.Mode.kSemiperiod;
+        else if (mode == FRC.CounterMode.kTwoPulse) counterMode = DirectCounter.Mode.kTwoPulse;
```
Comment (col6y) [inline roboRIO/src/ccre/frc/DirectFRCImplementation.java]: ... you know that you can switch on enums, right? Also PLEASE format your code.

File: roboRIO/src/ccre/frc/DirectFRCImplementation.java [hunk 18]
```diff
@@ -405,6 +405,35 @@ protected float apply() {
             }
         };
     }
+    
+    @Override
+    public FloatInput makeCounter(int channelUp, int channelDown, EventInput resetWhen, EventInput updateOn, FRC.CounterMode mode) {
+        if (channelUp < 0) throw new RuntimeException("Invalid up channel: " + channelUp);
```
Comment (col6y) [inline roboRIO/src/ccre/frc/DirectFRCImplementation.java]: Check the maximum too. And check `channelDown` too, at least as much as you can.

File: roboRIO/src/ccre/frc/DirectFRCImplementation.java [hunk 19]
```diff
@@ -405,6 +405,35 @@ protected float apply() {
...
+        if (mode == FRC.CounterMode.kExternalDirection) counterMode = DirectCounter.Mode.kExternalDirection;
+        else if (mode == FRC.CounterMode.kPulseLength) counterMode = DirectCounter.Mode.kPulseLength;
+        else if (mode == FRC.CounterMode.kSemiperiod) counterMode = DirectCounter.Mode.kSemiperiod;
+        else if (mode == FRC.CounterMode.kTwoPulse) counterMode = DirectCounter.Mode.kTwoPulse;
+        else throw new RuntimeException("Invalid mode: " + mode);
+
+        
+        ByteBuffer counter = channelDown >= 0 ? 
+                DirectCounter.init(channelUp, channelDown, counterMode)
+                : DirectCounter.init(channelUp, counterMode);
```
Comment (col6y) [inline roboRIO/src/ccre/frc/DirectFRCImplementation.java]: Couldn't you have this sentinel just be passed into `init`?

File: roboRIO/src/log-0 [hunk 20]
```diff
@@ -0,0 +1,6 @@
+Logging began at Wed Sep 30 16:55:15 PDT 2015 [0]
+[4 FINE] (ConnectionReceiverThread.java:58) About to listen on 1540
+[504 INFO] (DeviceListMain.java:99) Starting application: robot.RobotTemplate
+[509 INFO] (RobotTemplate.java:26) hello world
+[509 INFO] (DeviceListMain.java:102) Hello, robot.RobotTemplate!
+Logging continues at Wed Sep 30 16:55:25 PDT 2015
```
Comment (col6y) [inline roboRIO/src/log-0]: Don't commit logfiles!

File: CommonChickenRuntimeEngine/src/ccre/frc/FRC.java [hunk 21]
```diff
@@ -70,6 +70,24 @@
      */
     public static final int POWER_CHANNEL_6V = 3;
 
+    public static final int K_TWO_PULSE = 0;
+
+    public static final int K_SEMIPERIOD = 1;
+
+    public static final int K_PULSE_LENGTH = 2;
+
+    public static final int K_EXTERNAL_DIRECTION = 3;
```
Comment (col6y) [inline CommonChickenRuntimeEngine/src/ccre/frc/FRC.java]: Document these!

File: CommonChickenRuntimeEngine/src/ccre/frc/FRC.java [hunk 22]
```diff
@@ -725,6 +743,49 @@ public static FloatInput makeEncoder(int aChannel, int bChannel, boolean reverse
...
+     * @param channel The channel for the counter.
+     * @param resetWhen If provided, the Counter's value will be reset when this
+     * event is produced.
+     * @param updateOn Updates the FloatInput when this event is produced.
+     * @return the Counter, reporting Counter ticks.
+     */
+    public static FloatInput makeCounter(int upChannel, int downChannel, EventInput resetWhen, EventInput updateOn) {
+        return impl.makeCounter(upChannel, downChannel, resetWhen, updateOn, K_TWO_PULSE);
+    }
+
```
Comment (col6y) [inline CommonChickenRuntimeEngine/src/ccre/frc/FRC.java]: Where's the version with a mode but no updateOn? You'll want all combinations of the arguments that you have chosen to be optional.

File: CommonChickenRuntimeEngine/src/ccre/frc/FRC.java [hunk 23]
```diff
@@ -70,6 +70,24 @@
...
+    public static final int K_SEMIPERIOD = 1;
+
+    public static final int K_PULSE_LENGTH = 2;
+
+    public static final int K_EXTERNAL_DIRECTION = 3;
+    
+    /**
+     * Number of digital pins on a roboRIO.
+     */
+    public static final int K_DIGITAL_PINS = 26;
```
Comment (col6y) [inline CommonChickenRuntimeEngine/src/ccre/frc/FRC.java]: Do we need this here? We don't have it for anything else. Perhaps it should be in a differen tclass?

File: CommonChickenRuntimeEngine/src/ccre/frc/FRC.java [hunk 24]
```diff
@@ -70,6 +70,24 @@
...
+
+    /**
+     * For unused channels. Could be used for anything.
+     */
+    public static final int UNUSED = -1;
```
Comment (col6y) [inline CommonChickenRuntimeEngine/src/ccre/frc/FRC.java]: This makes it sound like you can put `UNUSED` in the place of any port number at all. Be more clear.

File: CommonChickenRuntimeEngine/src/ccre/frc/FRCImplementation.java [hunk 25]
```diff
@@ -201,6 +201,19 @@
...
     /**
+     * Creates a reference to a Counter on the specified port. When the channel
+     * is pulsed the counter will count up.
+     * 
+     * @param upChannel The DIO channel that will count up
+     * @param downChannel The DIO channel that will count down
+     * @param resetWhen If provided, the Counter's value will be reset when this
+     * event is produced.
+     * @param updateOn Updates the FloatInput when this event is produced.
+     * @return the Counter, reporting Counter ticks.
```
Comment (col6y) [inline CommonChickenRuntimeEngine/src/ccre/frc/FRCImplementation.java]: `@return`ing channels is usually documented in the CCRE as `@return a <X> that represents <Y>.` Look at other documentation for inspiration, because this reads slightly differently.

File: Emulator/src/ccre/frc/DeviceBasedImplementation.java [hunk 26]
```diff
@@ -496,4 +496,9 @@ public FloatInput addFloat(String name) {
     public EventInput getOnInitComplete() {
         return onInitComplete;
     }
+
+    @Override
+    public FloatInput makeCounter(int upChannel, int downChannel, EventInput resetWhen, EventInput updateOn, int mode) {
+        throw new UnsupportedOperationException("Not implemented.");
+    }
```
Comment (col6y) [inline Emulator/src/ccre/frc/DeviceBasedImplementation.java]: You need to implement this.

File: roboRIO/src/ccre/frc/DirectCounter.java [hunk 27]
```diff
@@ -0,0 +1,81 @@
...
+
+        if (channelUp != FRC.UNUSED)
+            setUpSource(counter, channelUp);
+        if (channelDown != FRC.UNUSED)
+            setDownSource(counter, channelDown);
+
+        // as done in WPILib, would be reversed if the counter should
+        // count in reverse direction
+        setUpSourceEdge(counter, true, false);
+        setDownSourceEdge(counter, false, true);
```
Comment (col6y) [inline roboRIO/src/ccre/frc/DirectCounter.java]: Explain what the arguments actually mean.

File: roboRIO/src/ccre/frc/DirectCounter.java [hunk 28]
```diff
@@ -0,0 +1,81 @@
...
+    public static ByteBuffer init(int channelUp, int channelDown, int mode) {
+        IntBuffer status = Common.getCheckBuffer();
+        IntBuffer index = Common.allocateInt();
+        ByteBuffer counter = CounterJNI.initializeCounter(mode, index, status);
+        Common.check(status);
+
+        if (channelUp != FRC.UNUSED)
+            setUpSource(counter, channelUp);
+        if (channelDown != FRC.UNUSED)
+            setDownSource(counter, channelDown);
```
Comment (col6y) [inline roboRIO/src/ccre/frc/DirectCounter.java]: Do you want to allow a counter with two UNUSED arguments? At the very least, I recommend that it shouldn't be allowed at this API layer. Perhaps a higher layer could check and send a `Logger.warning` about that, and then just return `FloatInput.always(0)`. You don't want to waste your counters.

File: roboRIO/src/ccre/frc/DirectCounter.java [hunk 29]
```diff
@@ -0,0 +1,81 @@
...
+        setUpSourceEdge(counter, true, false);
+        setDownSourceEdge(counter, false, true);
+
+        return counter;
+    }
+
+    public static void setUpSource(ByteBuffer counter, int channel) {
+        IntBuffer status = Common.getCheckBuffer();
+        if (DirectDigital.getDigitalSource(channel) == null)
+            DirectDigital.init(channel, true);
```
Comment (col6y) [inline roboRIO/src/ccre/frc/DirectCounter.java]: I recommend not auto-allocating at this API layer because it ends up with hard-to-predict behaviours when used in conjunction with freeing.

File: roboRIO/src/ccre/frc/DirectCounter.java [hunk 30]
```diff
@@ -0,0 +1,81 @@
...
+        if (!DirectDigital.isDigitalSourceInput(channel))
+            throw new RuntimeException("Cannot set up source as a digital output");
+        CounterJNI.setCounterUpSource(counter, channel, (byte) 0, status);
```
Comment (col6y) [inline roboRIO/src/ccre/frc/DirectCounter.java]: Please format this file according to the CCRE guidelines.

Comment (col6y) [inline roboRIO/src/ccre/frc/DirectCounter.java]: Also, specify what the `(byte) 0` means.

File: roboRIO/src/ccre/frc/DirectCounter.java [hunk 31]
```diff
@@ -0,0 +1,81 @@
+package ccre.frc;
+
+import java.nio.ByteBuffer;
+import java.nio.IntBuffer;
+
+import edu.wpi.first.wpilibj.hal.CounterJNI;
+
+public class DirectCounter {
+
+    public static ByteBuffer init(int channelUp, int channelDown, int mode) {
```
Comment (col6y) [inline roboRIO/src/ccre/frc/DirectCounter.java]: Where's your `free` method?

File: roboRIO/src/ccre/frc/DirectFRCImplementation.java [hunk 32]
```diff
@@ -99,7 +99,8 @@ private static void runMain() {
         DirectFRCImplementation robot = new DirectFRCImplementation();
         FRCImplementationHolder.setImplementation(robot);
         Cluck.setupServer(1540);// Cluck de-facto off-FMS port.
-        Cluck.setupServer(1735);// SmartDashboard port, since it's unused with the CCRE
+        Cluck.setupServer(1735);// SmartDashboard port, since it's unused with
+                                // the CCRE
```
Comment (col6y) [inline roboRIO/src/ccre/frc/DirectFRCImplementation.java]: Don't change this. I know it's autoformatted, but leave this at the previous version.

File: roboRIO/src/ccre/frc/DirectFRCImplementation.java [hunk 33]
```diff
@@ -407,6 +408,30 @@ protected float apply() {
     }
 
     @Override
+    public FloatInput makeCounter(int channelUp, int channelDown, EventInput resetWhen, EventInput updateOn, int mode) {
+        // unused happens to be -1, but this makes more sense than comparing
```
Comment (col6y) [inline roboRIO/src/ccre/frc/DirectFRCImplementation.java]: This is where you could check for both being UNUSED.

File: roboRIO/src/ccre/frc/DirectFRCImplementation.java [hunk 34]
```diff
@@ -502,7 +527,8 @@ public EventInput getDuringDisabled() {
 
     private synchronized ByteBuffer getPCMCompressor() {
         if (pcmCompressor == null) {
-            pcmCompressor = DirectCompressor.init(0);// TODO: Provide all PCM ids
+            pcmCompressor = DirectCompressor.init(0);// TODO: Provide all PCM
+                                                     // ids
```
Comment (col6y) [inline roboRIO/src/ccre/frc/DirectFRCImplementation.java]: Move that comment to its own line to avoid the formatting mess.

**Comment by col6y** [comment]
@Ja-ake my comments on this version are complete.

File: CommonChickenRuntimeEngine/src/ccre/frc/FRC.java [hunk 35]
```diff
@@ -71,6 +71,39 @@
...
+     * Pulse length mode for a counter. Uses the pulse width to determine if the
+     * counter should count up or down.
+     */
+    public static final int K_PULSE_LENGTH = 2;
+
+    /**
+     * External direction mode for a counter. Uses one source to count up and
+     * uses one source to determine direction.
+     */
+    public static final int K_EXTERNAL_DIRECTION = 3;
```
Comment (col6y) [inline CommonChickenRuntimeEngine/src/ccre/frc/FRC.java]: Hmm... should have caught this earlier, but perhaps `K_` isn't a good prefix since we don't do that anywhere else in the CCRE.

File: CommonChickenRuntimeEngine/src/ccre/frc/FRC.java [hunk 36]
```diff
@@ -71,6 +71,39 @@
...
+
+    /**
+     * For unused channels for counters.
+     * 
+     */
+    // TODO: update Javadoc if anything else uses this
+    public static final int UNUSED = -1;
```
Comment (col6y) [inline CommonChickenRuntimeEngine/src/ccre/frc/FRC.java]: That's not a valid TODO - it's a note for the future, not something that should be done now.
Also, please avoid extraneous blank lines in the Javadoc.

File: CommonChickenRuntimeEngine/src/ccre/frc/FRC.java [hunk 37]
```diff
@@ -717,7 +750,7 @@ public static FloatInput makeEncoder(int aChannel, int bChannel, boolean reverse
      * @param resetWhen If provided, the Encoder's value will be reset when this
      * event is produced.
      * @param updateOn when to update the sensor value.
-     * @return the Encoder, reporting encoder ticks.
+     * @return A float input that represents the count of an encoder
```
Comment (col6y) [inline CommonChickenRuntimeEngine/src/ccre/frc/FRC.java]: You don't have to fix this, but for posterity, I usually put "FloatInput", not "float input" in the phrase "A FloatInput that represents X." Sorry for not catching that the first time, so it doesn't matter if you fix this.

File: CommonChickenRuntimeEngine/src/ccre/frc/FRC.java [hunk 38]
```diff
@@ -725,6 +758,63 @@ public static FloatInput makeEncoder(int aChannel, int bChannel, boolean reverse
     }
 
     /**
+     * Creates a reference to a Counter on the specified ports with the
+     * specified number of average bits. Will count up when up channel is
+     * pulsed, and down when down channel is pulsed.
+     * 
+     * @param channel The channel for the counter.
```
Comment (col6y) [inline CommonChickenRuntimeEngine/src/ccre/frc/FRC.java]: Specify the delta per tick?

File: CommonChickenRuntimeEngine/src/ccre/frc/FRC.java [hunk 39]
```diff
@@ -725,6 +758,63 @@ public static FloatInput makeEncoder(int aChannel, int bChannel, boolean reverse
     }
 
     /**
+     * Creates a reference to a Counter on the specified ports with the
+     * specified number of average bits. Will count up when up channel is
+     * pulsed, and down when down channel is pulsed.
+     *
```
Comment (col6y) [inline CommonChickenRuntimeEngine/src/ccre/frc/FRC.java]: Specify delta per tick?

File: CommonChickenRuntimeEngine/src/ccre/frc/FRCImplementation.java (hunk 25 above)
Comment (col6y) [inline CommonChickenRuntimeEngine/src/ccre/frc/FRCImplementation.java]: Please update this here too.

File: Emulator/src/ccre/frc/DeviceBasedImplementation.java [hunk 40]
```diff
@@ -254,6 +254,12 @@ public FloatInput makeEncoder(int aChannel, int bChannel, boolean reverse, Event
         return panel.add(new SpinDevice("Encoder " + aChannel + ":" + bChannel + (reverse ? " (REVERSED)" : ""), resetWhen)).asInput();
     }
 
+    @Override
+    public FloatInput makeCounter(int aChannel, int bChannel, EventInput resetWhen, EventInput updateOn, int mode) {
+        return panel.add(new SpinDevice("Counter " + (aChannel == -1 ? "UNUSED" : ("" + aChannel)) + ":" 
+                        + (aChannel == -1 ? "UNUSED" : ("" + aChannel)), resetWhen)).asInput();
```
Comment (col6y) [inline Emulator/src/ccre/frc/DeviceBasedImplementation.java]: Don't you mean `bChannel`? Also, `String.valueOf` or `Integer.toString` are preferred to `"" + ` in the CCRE.

File: roboRIO/src/ccre/frc/DirectCounter.java [hunk 41]
```diff
@@ -0,0 +1,105 @@
...
+        if (DirectDigital.getDigitalSource(channel) == null) {
+            throw new RuntimeException("Digital source has not been allocated yet");
+        }
+        if (!DirectDigital.isDigitalSourceInput(channel)) {
+            throw new RuntimeException("Cannot set up source as a digital output");
+        }
+
+        // the third argument is a c++ bool that represents if the counter
+        // source is an analog input
+        CounterJNI.setCounterUpSource(counter, channel, (byte) 0, status);
```
Comment (col6y) [inline roboRIO/src/ccre/frc/DirectCounter.java]: Optional, but perhaps better way to do this: have byte constants for 0 and 1 to represent that argument's possible values.

File: roboRIO/src/ccre/frc/DirectDigital.java [hunk 42]
```diff
@@ -34,6 +34,11 @@
 
     static final byte WATCH_ASYNCHRONOUS = 0, WATCH_SYNCHRONOUS = 1;
     static final byte TRIGGER_DIGITAL = 0, TRIGGER_ANALOG = 1;
+    
+    /**
+     * Number of digital pins on a roboRIO.
+     */
+    static final int K_DIGITAL_PINS = 26;
```
Comment (col6y) [inline roboRIO/src/ccre/frc/DirectDigital.java]: Same `K_` prefix thing applies here as above.

File: roboRIO/src/ccre/frc/DirectFRCImplementation.java [hunk 43]
```diff
@@ -407,6 +407,35 @@ protected float apply() {
     }
 
     @Override
+    public FloatInput makeCounter(int channelUp, int channelDown, EventInput resetWhen, EventInput updateOn, int mode) {
+        // unused happens to be -1, but this makes more sense than comparing
+        // with -1
+        if (channelUp == FRC.UNUSED && channelDown == FRC.UNUSED) {
+            Logger.warning("Both channelUp and channelDown are unused.");
```
Comment (col6y) [inline roboRIO/src/ccre/frc/DirectFRCImplementation.java]: Perhaps add "on an instantiated Counter".

File: roboRIO/src/ccre/frc/DirectFRCImplementation.java [hunk 44]
```diff
@@ -407,6 +407,35 @@ protected float apply() {
...
+            return FloatInput.zero;
+        }
+        
+        if (channelUp != FRC.UNUSED && (channelUp < 0 || channelUp >= DirectDigital.K_DIGITAL_PINS))
+            throw new RuntimeException("Invalid up channel: " + channelUp);
+        if (channelDown != FRC.UNUSED && (channelDown < 0 || channelDown >= DirectDigital.K_DIGITAL_PINS))
+            throw new RuntimeException("Invalid down channel: " + channelDown);
```
Comment (col6y) [inline roboRIO/src/ccre/frc/DirectFRCImplementation.java]: Format this.

**Comment by col6y** [comment]
Almost there! Just a few nitpicks.

**Comment by Ja-ake** [comment]
Ready for rereview.

File: CommonChickenRuntimeEngine/src/ccre/frc/FRC.java [hunk 45]
```diff
@@ -725,6 +756,75 @@ public static FloatInput makeEncoder(int aChannel, int bChannel, boolean reverse
...
+     * @see FRC#TWO_PULSE_MODE
+     * 
+     * @param channel The channel for the counter.
+     * @param resetWhen If provided, the Counter's value will be reset when this
+     * event is produced.
+     * @return A FloatInput that represents the count of a counter
+     */
+    public static FloatInput makeCounter(int upChannel, int downChannel, EventInput resetWhen) {
+        return impl.makeCounter(upChannel, downChannel, resetWhen, sensorPeriodic, TWO_PULSE_MODE);
+    }
```
Comment (col6y) [inline CommonChickenRuntimeEngine/src/ccre/frc/FRC.java]: This all still say `makeCounter`. Sorry, I must have had you reset that. Hopefully this is a fast fix.

File: roboRIO/src/ccre/frc/DirectCounter.java [hunk 46]
```diff
@@ -0,0 +1,108 @@
...
+
+import java.nio.ByteBuffer;
+import java.nio.IntBuffer;
+
+import edu.wpi.first.wpilibj.hal.CounterJNI;
+
+public class DirectCounter {
+
+    public static byte ANALOG_INPUT = 1;
+    public static byte DIGITAL_INPUT = 0;
```
Comment (col6y) [inline roboRIO/src/ccre/frc/DirectCounter.java]: `final`. Otherwise, these aren't actually constants.

File: roboRIO/src/ccre/frc/DirectCounter.java [hunk 47]
```diff
@@ -0,0 +1,108 @@
...
+        Common.check(status);
+    }
+
+    public static void setUpSource(ByteBuffer counter, int channel) {
+        IntBuffer status = Common.getCheckBuffer();
+        if (DirectDigital.getDigitalSource(channel) == null) {
+            throw new RuntimeException("Digital source has not been allocated yet");
+        }
+        if (!DirectDigital.isDigitalSourceInput(channel)) {
+            throw new RuntimeException("Cannot set up source as a digital output");
```
Comment (col6y) [inline roboRIO/src/ccre/frc/DirectCounter.java]: This message is confusing and backwards.

File: roboRIO/src/ccre/frc/DirectCounter.java [hunk 48]
```diff
@@ -0,0 +1,108 @@
...
+        Common.check(status);
+    }
+
+    public static void setDownSource(ByteBuffer counter, int channel) {
+        IntBuffer status = Common.getCheckBuffer();
+        if (DirectDigital.getDigitalSource(channel) == null) {
+            throw new RuntimeException("Digital source has not been allocated yet");
+        }
+        if (!DirectDigital.isDigitalSourceInput(channel)) {
+            throw new RuntimeException("Cannot set up source as a digital output");
```
Comment (col6y) [inline roboRIO/src/ccre/frc/DirectCounter.java]: Same.

**Comment by col6y** [comment]
Four minor things.

File: CommonChickenRuntimeEngine/src/ccre/frc/FRC.java [hunk 49]
```diff
@@ -72,6 +72,37 @@
...
+     * External direction mode for a counter. Uses one source to count up and
+     * uses one source to determine direction.
+     */
+    public static final int EXTERNAL_DIRECTION_MODE = 3;
+
+    /**
+     * For unused channels for counters.
+     */
+    public static final int UNUSED = -1; 
+
```
Comment (Others) [inline CommonChickenRuntimeEngine/src/ccre/frc/FRC.java]: Shouldn't this be an enum?

File: Emulator/src/ccre/frc/DeviceBasedImplementation.java [hunk 50]
```diff
@@ -254,6 +254,12 @@ public FloatInput makeEncoder(int aChannel, int bChannel, boolean reverse, Event
         return panel.add(new SpinDevice("Encoder " + aChannel + ":" + bChannel + (reverse ? " (REVERSED)" : ""), resetWhen)).asInput();
     }
 
+    @Override
+    public FloatInput makeCounter(int aChannel, int bChannel, EventInput resetWhen, EventInput updateOn, int mode) {
+        return panel.add(new SpinDevice("Counter " + (aChannel == -1 ? "UNUSED" : ("" + aChannel)) + ":" 
+                        + (bChannel == -1 ? "UNUSED" : ("" + bChannel)), resetWhen)).asInput();
+    }
```
Comment (Others) [inline Emulator/src/ccre/frc/DeviceBasedImplementation.java]: Isn't -1 a magic number?

File: CommonChickenRuntimeEngine/src/ccre/frc/FRC.java (hunk 49 above)
Comment (Ja-ake) [inline CommonChickenRuntimeEngine/src/ccre/frc/FRC.java]: Colby wants these to be integers.

**Comment by Ja-ake** [comment]
In theory everything should be fixed.

**Comment by col6y** [comment]
I am realizing that you didn't add your copyright to the files you changed significantly. I'm doing that for you now.
